oracle/blockoracle.go

```diff
@@ -0,0 +1,88 @@
+package oracle
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+)
+
+type ActivationDb interface {
+	GetNodeAtxIds(node mesh.NodeId) ([]mesh.AtxId, error)
+	GetAtx(id mesh.AtxId) (*mesh.ActivationTx, error)
+}
+
+type MinerBlockOracle struct {
+	committeeSize  int32
+	layersPerEpoch uint16
+	activationDb   ActivationDb
+	beaconProvider *EpochBeaconProvider
+	vrfSigner      *crypto.VRFSigner
+	nodeID         mesh.NodeId
+
+	eligibilityProofs map[mesh.LayerID][]mesh.BlockEligibilityProof
+	proofsEpoch       uint64
+}
+
+func NewMinerBlockOracle(committeeSize int32, layersPerEpoch uint16, activationDb ActivationDb,
+	beaconProvider *EpochBeaconProvider, vrfSigner *crypto.VRFSigner, nodeId mesh.NodeId) *MinerBlockOracle {
+
+	return &MinerBlockOracle{
+		committeeSize:  committeeSize,
+		layersPerEpoch: layersPerEpoch,
+		activationDb:   activationDb,
+		beaconProvider: beaconProvider,
+		vrfSigner:      vrfSigner,
+		nodeID:         nodeId,
+		proofsEpoch:    ^uint64(0),
+	}
+}
+
+func (bo *MinerBlockOracle) BlockEligible(layerID mesh.LayerID) ([]mesh.BlockEligibilityProof, error) {
+
+	epochNumber := uint64(layerID) / uint64(bo.layersPerEpoch)
+
+	if bo.proofsEpoch != epochNumber {
+		err := bo.calcEligibilityProofs(epochNumber)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return bo.eligibilityProofs[layerID], nil
+}
+
+func (bo *MinerBlockOracle) calcEligibilityProofs(epochNumber uint64) error {
+	epochBeacon := bo.beaconProvider.GetBeacon(epochNumber)
+
+	atxIDs, err := bo.activationDb.GetNodeAtxIds(bo.nodeID)
+	if err != nil {
+		log.Error("getting previous atx id failed: %v", err)
+		return err
+	}
+	atxId := atxIDs[len(atxIDs)-1]
+	atx, err := bo.activationDb.GetAtx(atxId)
+	activeSetSize := atx.ActiveSetSize
+
+	numberOfEligibleBlocks := uint32(bo.committeeSize) * uint32(bo.layersPerEpoch) / activeSetSize
```

Comment:
I need to check if committee size is constant or varies somehow, I'll get back to you

Reply:
Go ahead and check, it can't hurt.

I've discussed this yesterday with @barakshani and @zalmen and we've agreed that while the *actual* committee size (or avg. size) will change, the value we're aiming for *won't* (and this is what `committeeSize` stands for here).

Reply:
@zalmen shouldn't it be num of active Ids in last epoch?

Reply:
I think you're confusing `committeeSize` with `activeSetSize`. The latter is calculated from the ATX's view.

`committeeSize` is the number of blocks we want to have per layer and it's configured to be 200 at the moment. Since we promise each miner to be eligible at least once per epoch, the actual avg. committee size can become bigger if we have more than 806,400 active mining identities.